### ISSUE TITLE
Get user avatar from ZeroMe

### ISSUE THREAD

**Issue by Plasmmer** [AnthyG/ThunderWave#23]

**Comment by Plasmmer** [comment]
@mkg20001, is it possible to achieve?

**Comment by mkg20001** [comment]
Yes, but one needs to download all the hubs and thunderwave needs a db

**Comment by AnthyG** [comment]
ThunderWave has a DB, but how would I download *all* (or at least, all needed) hubs?
Before, I already checked some source code of ZeroMe, so I may be able to reproduce the needed result.. 🤔

**Comment by AnthyG** [comment]
Just found this link: http://127.0.0.1:43110/1oranGeS2xsKZ4jVsu9SVttzgkYXu4k9v/data/users/14K7EydgyeP84L1NKaAHBZTPQCev8BbqCy/avatar.jpg?1492026832
Hint: *It shows my avatar, saved in OrangeHub!*

[This is a link](https://github.com/HelloZeroNet/ZeroMe/blob/master/js/User.coffee#L67) to the function, which returns the link for the avatar in ZeroMe, and the following is the compiled version of it:
```
User.prototype.getAvatarLink = function() {
      var cache_invalidation, ref;
      cache_invalidation = "";
      if (this.auth_address === ((ref = Page.user) != null ? ref.auth_address : void 0)) {
        cache_invalidation = "?" + Page.cache_time;
      }
      return "merged-ZeroMe/" + this.hub + "/data/users/" + this.auth_address + "/avatar." + this.row.avatar + cache_invalidation;
};
```

So now I just have to find a way to download the ZeroMe-database or every needed Hub.. 🤔💭

**Comment by AnthyG** [comment]
So we just need a function that gets the user's hub (and that of every other user who has sent a message on ThunderWave..)

**Comment by AnthyG** [comment]
Somewhat added with commit 7714e53e892eb97f5088b6f77b6c91d211084903

**Comment by AnthyG** [comment]
Next up for this is to add two options in the Settings-Page:
1. Upload Avatar (for ThunderWave image-source)
2. Set File URL (for external image-source)

**Comment by Plasmmer** [comment]
Would be better also getting it automatically from ZeroMe as default avatar, and the upload/URL as option.

**Comment by AnthyG** [comment]
Getting the Avatars automatically, would require to know, in which Hub the user is,.. I'm not sure, if I can do that, without loading all the hubs.

**Comment by Plasmmer** [comment]
I'll think about it.
And about the thunderwave.bit domain, just send me your namecoin address.

**Comment by mkg20001** [comment]
The hub of the user can be found in the user DB. The URL should be 1UDB...

_Sent from my Google Nexus 5X using [FastHub](https://play.google.com/store/apps/details?id=com.fastaccess.github)_

**Comment by AnthyG** [comment]
@mkg20001 Are *all* Users saved in there?
It seems to be a _huge_ data-dir (~6.1MB)

**Comment by mkg20001** [comment]
1UDbADib99KE9d3qZ87NqJF2QLTHmMkoV has 79mb
Every user using ZeroMe seeds it.
The hub can be found in 1UDb/data/userdb/1UserID/content.json in the user.hub field

**Comment by AnthyG** [comment]
Yeah, see that 👍

**Comment by AnthyG** [comment]
Well that's easy then 😆

**Comment by Plasmmer** [comment]
@mkg20001 @AnthyG you are great. Thanks for this.
Will be very nice to have ThunderWave getting the avatar automatically, without need to user upload it.

**Comment by AnthyG** [comment]
You can't believe, how hard it is for me, to wrap my head around the really easy `if this, then that`-method for this 😅 (It took me so long to create the original one, and now I have to rewrite it completely :/)

**Comment by AnthyG** [comment]
Ok, I believe I have covered all cases now.
Please read through the following, and maybe try and check if it is correct/ if anything is missing.

So e.g.:
User 1 (_u1_) views User 2 (_u2_)

---

_u1_ can decide on a range from 0 - 3:
- 0 = Try to get from ThunderWave, or ZeroMe, or generate
- 1 = Try to get from ThunderWave, or generate
- 2 = Try to get from ZeroMe, or generate
- 3 = generate

_u2_ can decide on a range from 0 - 2:
- 0 = generate
- 1 = ThunderWave
- 2 = ZeroMe

---

If _u2_ has set 0, then generate (no matter what _u1_ has set)

If _u2_ has set 1, and the avatar-file-name:
- If _u1_ has set 0, then use ThunderWave
- If _u1_ has set 1, then use ThunderWave
- If _u1_ has set 2, then generate
- If _u1_ has set 3, then generate

If _u2_ has set 1, without avatar-file-name:
- If _u1_ has set 0, then check if _u2_ has ZeroMe:
- - If yes, then use ZeroMe
- - If no, then generate
- If _u1_ has set 1, then generate
- If _u1_ has set 2, then generate
- If _u1_ has set 3, then generate

If _u2_ has set 2:
- If _u1_ has set 0, then check if _u2_ has ZeroMe:
- - If yes, then use ZeroMe
- - If no, then check if _u2_ has set avatar-file-name:
- - - If yes, then use ThunderWave
- - - If no, then generate
- If _u1_ has set 1, then generate
- If _u1_ has set 2, then check if _u2_ has ZeroMe:
- - If yes, then use ZeroMe
- - If no, then generate
- If _u1_ has 3, then generate

**Comment by AnthyG** [comment]
I'm considering this as closed with commit 050baacbba2b4f967cce028528b45019db6c0a3b.
